Kotlin-test: Add assertIs assertion

benasher44 opened the pull request:
Kotlin 1.4 supports using the type parameter in its contract. This will allow code like this in tests:
```kotlin
val s: Any = "test"
assertIs<String>(s)
// can now print s.length because of contract in assertIs
println("${s.length}")
```

[KT-45296](https://youtrack.jetbrains.com/issue/KT-45296)

benasher44 commented:
The main value of this over `assertTrue(value is T)` is the built-in nice error message that tells you what type the value ended up being upon failure

ilya-g commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Assertions.kt:
```diff
@@ -72,6 +72,21 @@ fun <@OnlyInputTypes T> assertNotSame(illegal: T, actual: T, message: String? =
 }
 
+/** Asserts that [value] is [T], with an optional [message]. */
+@SinceKotlin("1.4")
```
Now it's time for `SinceKotlin("1.5")`  : )

ilya-g commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Assertions.kt:
```diff
@@ -72,6 +72,21 @@ fun <@OnlyInputTypes T> assertNotSame(illegal: T, actual: T, message: String? =
+/** Asserts that [value] is [T], with an optional [message]. */
+@SinceKotlin("1.4")
+@InlineOnly
+inline fun <reified T: Any> assertIs(value: Any, message: String? = null) {
```
It would be valuable to support it for nullable type and value.

ilya-g commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Assertions.kt:
```diff
@@ -72,6 +72,21 @@ fun <@OnlyInputTypes T> assertNotSame(illegal: T, actual: T, message: String? =
+@InlineOnly
+inline fun <reified T: Any> assertIs(value: Any, message: String? = null) {
+    contract { returns() implies (value is T) }
+    asserter.assertTrue({ messagePrefix(message) + "Expected value to be of type <${T::class}>, actual <${value::class}>." }, value is T)
```
`typeOf<T>()` provides more precise information about the type, especially its nullability.

ilya-g commented on libraries/kotlin.test/common/src/test/kotlin/kotlin/test/tests/BasicAssertionsTest.kt:
```diff
@@ -205,6 +205,28 @@ class BasicAssertionsTest {
+    
+    @Test
+    fun testAssertIsFails() {
+        checkFailedAssertion { assertIs<Int>("test") }
```
Could you inspect the assertion returned by `checkFailedAssertion` and check that its message contains the information we expect?

ilya-g commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Assertions.kt:
```diff
@@ -72,6 +72,21 @@ fun <@OnlyInputTypes T> assertNotSame(illegal: T, actual: T, message: String? =
     asserter.assertNotSame(message, illegal, actual)
 }
 
+/** Asserts that [value] is [T], with an optional [message]. */
```
```suggestion
/** Asserts that [value] is of type [T], with an optional [message]. */
```

Also makes sense to add a note that due to type erasure the type check may be partial, e.g. `assertIs<List<String>>(value)` only checks for the class being `List` and doesn't check the type of elements because it's erased.

ilya-g commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Assertions.kt:
```diff
@@ -72,6 +72,21 @@ fun <@OnlyInputTypes T> assertNotSame(illegal: T, actual: T, message: String? =
+    asserter.assertTrue({ messagePrefix(message) + "Expected value to be of type <${T::class}>, actual <${value::class}>." }, value is T)
+}
+
+/** Asserts that [value] !is [T], with an optional [message]. */
```
```suggestion
/** Asserts that [value] is not of type [T], with an optional [message]. */
```

ilya-g commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Assertions.kt:
```diff
@@ -72,6 +72,21 @@ fun <@OnlyInputTypes T> assertNotSame(illegal: T, actual: T, message: String? =
+@SinceKotlin("1.4")
+@InlineOnly
+inline fun <reified T: Any> assertNotIs(value: Any, message: String? = null) {
+    asserter.assertFalse({ messagePrefix(message) + "Expected value to not be of type <${T::class}>" }, value is T)
```
`messagePrefix` is internal and needs a `@PublisedApi` annotation to be used in an inline function.

benasher44 replied:
Would you recommend that? I could also add an internal non-inline helper function that calls it. What do you think?

benasher44 replied:
Oh never mind. That's actually the same problem. Do you think `@PublishedApi` is the way to go?

ilya-g replied:
Yes, I think it's ok to expose this function (`messagePrefix`) to public binary API.

benasher44 commented:
Thanks for the feedback! I should be able to get to it in the next week.

benasher44 commented:
Fixed some of the initial issues. Left one question, which may have impact on the other feedback, and then I'll fix those final ones too.

ilya-g commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Assertions.kt:
```diff
@@ -72,6 +73,31 @@ fun <@OnlyInputTypes T> assertNotSame(illegal: T, actual: T, message: String? =
+@InlineOnly
+inline fun <reified T> assertIs(value: Any?, message: String? = null) {
+    contract { returns() implies (value is T) }
+    asserter.assertTrue({ messagePrefix(message) + "Expected value to be of type <${typeOf<T>()}>, actual <${value::class}>." }, value is T)
```
Please ensure that the library compiles and its tests pass. `coreLibsTest` gradle task can be used for that.
In particular, `value::class` won't compile for nullable `value` because `::class` requires a non nullable LHS argument. We can use `value?.let { it::class }` instead.

benasher44 replied:
Yes of course. Sorry I wasn't done yet— just wanted to get the initial improvements in

benasher44 commented on libraries/kotlin.test/common/src/main/kotlin/kotlin/test/Utils.kt:
```diff
@@ -5,8 +5,9 @@
 @PublishedApi // required to get stable name as it's called from box tests
-internal fun overrideAsserter(value: Asserter?): Asserter? = _asserter.also { _asserter = value }
\ No newline at end of file
+internal fun overrideAsserter(value: Asserter?): Asserter? = _asserter.also { _asserter = value }
```
😫 will fix

benasher44 commented:
Okay fixed final issues, cleaned up some white space things, and squashed it all into one neat commit. I tested locally in a scratch file, and I think everything should be compiling now 🤞

ilya-g commented on libraries/kotlin.test/common/src/test/kotlin/kotlin/test/tests/BasicAssertionsTest.kt:
```diff
@@ -205,11 +205,35 @@ class BasicAssertionsTest {
-internal fun checkFailedAssertion(assertion: () -> Unit): AssertionError {
-    return assertFailsWith<AssertionError> { withDefaultAsserter(assertion) }
+internal fun checkFailedAssertion(message: String? = null, assertion: () -> Unit): AssertionError {
+    return assertFailsWith(message)<AssertionError> { withDefaultAsserter(assertion) }
```
This is a message for failed `assertFailsWith`, not the message to check in the successfully caught exception.
Use the `AssertionError` returned from `checkFailedAssertion` to verify its properties in particular tests.

benasher44 replied:
🤦‍♂️ thanks!

ilya-g commented on libraries/kotlin.test/common/src/test/kotlin/kotlin/test/tests/BasicAssertionsTest.kt:
```diff
@@ -205,11 +205,35 @@ class BasicAssertionsTest {
+
+    @Test
+    fun testAssertIsFails() {
+        val message = "Expected value to be of type <kotlin.Int>, actual <class kotlin.String>."
```
Note that different platforms can have different `toString` representations of `KType` and `KClass`. Better to check that the message contains essential information rather than to expect it in exact form.

ilya-g commented:
Thanks for the PR, I've merged it manually with some changes: https://github.com/JetBrains/kotlin/compare/e66eeefe~2...e66eeefe

In particular we've decided to rename `assertNotIs` to `assertIsNot` because the latter reads better.

benasher44 commented:
Awesome thanks!!
